src/commands/login.rs

```diff
@@ -98,54 +102,218 @@ where
     }
 }
 
+async fn open_that(url: Url) -> bool {
+    tokio::task::spawn_blocking(move || open::that(url.as_str()))
+        .await
+        .is_ok()
+}
+
 async fn get_oauth_code(
     global: &GlobalArgs,
     client_id: &str,
     progress: &ProgressBar,
+    interactive: bool,
 ) -> Result<Option<(Url, String)>> {
+    let (tx, rx) = oneshot::channel();
+    let state = ServerState::new(tx);
+
+    let router = Router::<ServerState<LoginResponse>>::new()
+        .route("/", get(login_callback))
+        .with_state(state);
     let listener = TcpListener::bind("127.0.0.1:0").await.map_err(|e| {
         error::user(
             &format!("Could not bind to any port for OAuth callback: {e:?}"),
             "Make sure you have permission to bind to a network port",
         )
     })?;
     let port = listener.local_addr()?.port();
-    let redirect_uri = Url::parse(&format!("http://localhost:{port}"))?;
+    let http = axum::serve(listener, router)
+        .with_graceful_shutdown(shutdown_signal())
+        .into_future();
+
+    tokio::time::sleep(std::time::Duration::from_millis(500)).await;
     let session = BASE64_URL_SAFE_NO_PAD.encode(rand::random::<[u8; 16]>());
-    let (tx, rx) = oneshot::channel();
-    let state = ServerState::new(tx);
-    let router = Router::<ServerState<LoginResponse>>::new()
-        .route("/", get(login_callback))
-        .with_state(state);
+    let redirect_uri = Url::parse(&format!("http://localhost:{port}"))?;
     let authorize_url = global.authorize_url(client_id, &redirect_uri, &session)?;
-    let cloned_progress = progress.clone();
-    tokio::spawn(async move {
-        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
-        cloned_progress.set_message(format!("Opening {authorize_url}..."));
-        if open::that(authorize_url.as_str()).is_err() {
-            cloned_progress.set_message(format!(
-                "Failed to open browser, please open {authorize_url} manually"
-            ));
-        }
-        cloned_progress.set_message("Waiting for browser response...");
-    });
+
+    if !interactive {
+        progress.suspend(|| {
+            // NOTE: suspending here here instead of just `progress.println(...)`
+            // because indicatif will cut extra characters instead of wrapping
+            println!("Please navigate to this url (it should be done automatically):");
```

Comment:
I would just say `Opening {authorize_url} (it should open automatically)` that way people don't thing they have to do anything. Maybe also add a helpful message for people that don't have access to a browser to use `--interactive` or whatever

Reply:
> I would just say Opening {authorize_url} (it should open automatically) that way people don't thing they have to do anything.

`authorize_url` is a very long string, so I decided to put on its line in order to clearly show `(it should open automatically)`

> Maybe also add a helpful message for people that don't have access to a browser to use --interactive or whatever

Great idea 🙂 added it!